File: apps/api/core/roles.py
ROLE_HIERARCHY = {"guest": 0, "resident": 1, "driver": 2, "admin": 3}

REGISTERABLE_ROLES = {"resident", "driver"}


def role_satisfies(user_role: str, required_role: str) -> bool:
    return ROLE_HIERARCHY[user_role] >= ROLE_HIERARCHY[required_role]

File: packages/constants/src/roles.ts
export const ROLES = ["guest", "resident", "driver", "admin"] as const;
export type Role = (typeof ROLES)[number];

export const ROLE_HIERARCHY: Record<Role, number> = {
  guest: 0,
  resident: 1,
  driver: 2,
  admin: 3,
};

export function roleSatisfies(userRole: Role, requiredRole: Role): boolean {
  return ROLE_HIERARCHY[userRole] >= ROLE_HIERARCHY[requiredRole];
}
